WIP OPRs, Syncing, Config, more

WhoSoup opened the pull request:
I'm going to use this as a "work in progress" branch so that you can review the changes. Doing single-feature branches doesn't make a lot of sense yet 

Changes:

## Flags
Added --server (-s) and --wallet (-w) as flags to specify different endpoints. I'm not sure if your emyrk/factom branch is outdated but it doesn't match up with the FAT/factom branch when it comes to creating a new client. 

## Persistence

### SQL
I changed the table create queries to include "IF NOT EXISTS" so we can get rid of the awkward double error check.

### Metadata

I created a metadata table that stores key-value settings. One of those settings is "synced", which is used to save the last synced height. When restarted, the node will pick up where it left off. If the setting doesn't exist, it starts from the hardcoded number. 

### Grade & Rates

This is mostly copying code over from my FAT branch. I changed the Grade() function to pull the previous winners from the database and persist the new winners to the database. We'll have to extend this to include additional information but it works for now.

Emyrk commented:
I know Adam made some updates after my fork. I'll bring them into my fork when I get the chance today.

WhoSoup commented:
More changes:

* following #3 I tore out all references to Network, hardcoded the opr chain and burn addresses
* cleaned up a bunch of comments, removed a couple of TODOs , fixed a couple of lint issues
* consolidated the boilerplate table creation functions into a big one

Emyrk commented on node/opr.go:
```diff
@@ -1,33 +1,37 @@
-	prev := d.Pegnet.FetchPreviousBlock()
-	if prev != nil {
-		prevWinners = prev.WinnersShortHashes()
+	prev, err := d.Pegnet.SelectGrade(ctx, block.Height-1)
```
A height could have no graded block. It is possible to skip a height, or even have an invalid eblock. We should really select the last valid graded block.

WhoSoup replied:
my first thought was to just make it write a list of unbroken winners regardless of height (so instead of empty heights it would repeat entries) but i don't think that's necessary. changing it to a query that selects the most recent entry instead

WhoSoup replied:
done

Emyrk commented on node/opr.go:
```diff
@@ -1,33 +1,37 @@
+	prev, err := d.Pegnet.SelectGrade(ctx, block.Height-1)
+	// assume that error means it's below genesis for now
+	if err != nil {
+		log.WithError(err).Debug("failed to get previous winners")
```
We can check the error for `Notfound`, otherwise, actually return an error.

This will fork a node later if we have a db issue.

WhoSoup replied:
done

Emyrk commented on node/pegnet/grading.go:
```diff
@@ -0,0 +1,65 @@
+	"winners" BLOB,
+	
+	UNIQUE("height")
+);
```
I think it'd be nice to have some more context around the eblock keymr and sequence.

If we only index by height, we can only traverse by height, and there will be gaps. If we also store
```
Eblock keymr
Eblock Seq
Prev KeyMr
```

We can traverse chains cleanly. I could see value in this indexing. I think it would also be a nice index to store the previous keymr/height/seq (choose one, idc) of the last eblock with a valid graded block. Since not all eblocks necesarily have a set of winners.

WhoSoup replied:
traversing by height is the same thing as traversing by prevkeymr and sequence but way better than prevkey. sequence has some merits but we're not really doing anything by sequence in here. storing the previous index is pointless in an sql environment, you only do that in key-value databases


i think we should store the eblock keymr for lookup purposes but i don't think the others are necessary

WhoSoup replied:
i ended up just adding them since it didn't really cost anything

Emyrk commented on node/pegnet/grading.go:
```diff
@@ -0,0 +1,65 @@
+}
+
+func (p *Pegnet) SelectGrade(ctx context.Context, height uint32) ([]string, error) {
+	// this assumes that all previous heights have rows
```
What do we do if the previous height does not? Like no oprs, or no winners?

WhoSoup replied:
duplicate of https://github.com/pegnet/pegnetd/pull/2#discussion_r330260181

Emyrk commented on node/pegnet/metadata.go:
```diff
@@ -0,0 +1,35 @@
+func (p *Pegnet) InsertSynced(ctx context.Context, height uint32) error {
+	b := make([]byte, 4)
+	binary.BigEndian.PutUint32(b, height)
+	_, err := p.DB.ExecContext(ctx, "REPLACE INTO pn_metadata (name, value) VALUES ($1, $2)", "synced", b)
```
I think we should just make a `Synced struct` and json marshal it to the db. I could see us wanting to extend this later, and being able to add fields would make backward compatibility easier. 

We should eat the extra cost of storage now, to give us easier extendibility later.

WhoSoup replied:
the metadata supports a bunch of key-value options. if we want extendability, we can just read the entire "metadata" table into a map[string]string

doing it this way allows us to use "synced" in queries

WhoSoup replied:
since we can't query it anyway might as well make it a struct.

i moved the BlockSync struct to pegnet.BlockSync, which atm only contains Synced. if we add more records we should specify the "omitempty" tag for backward compatibility

Emyrk replied:
I was unaware you were trying to make it a queryable field. 

As long as it's a blob, json might remove headaches later :+1:

Emyrk commented on node/sync.go:
```diff
@@ -83,37 +83,24 @@ func (d *Pegnetd) SyncBlock(ctx context.Context, height uint32) error {
-					break EntrySyncLoop
-				}
+			if err = eblock.GetEntries(d.FactomClient); err != nil {
+				return err
```
TIL

WhoSoup replied:
yeah the library is inconsistent. you can do eblock.GetEntries and it will get both the eblock and entries, but dblock.geteblocks doesn't work the same

Emyrk commented on node/sync.go:
```diff
@@ -124,12 +111,25 @@ EntrySyncLoop: // Syncs all eblocks we care about and their entries
+	// TODO: Handle converts/txs
+
+	// update synced state
+	err = d.Pegnet.InsertSynced(ctx, height)
```
We should use a sql tx to make sure all db writes are all or nothing per block

WhoSoup replied:
yes i planned that for a future pr

WhoSoup replied:
i was getting a lot of corrupted databases by just ctrl+c'ing out of the node and trying to restart it, so i ended up wrapping it in a tx

Emyrk replied:
gotcha :+1:

WhoSoup commented:
Changes:

* Using the most recent set of winners instead of height-1
* Debug notices improved for starting with a clean db/retrieving sync height, now only throws errors if the data is corrupt
* I extended pn_grade to include a lot more metadata and created a pn_winners that saves the metadata of all of the **graded** oprs (50 per height)

WhoSoup commented:
SyncBlock is now wrapped in a transaction.

all function called inside of SyncBlock should only execute queries on the transaction, not to the database itself

Emyrk commented on node/sync.go:
```diff
@@ -43,19 +40,49 @@ OuterSyncLoop:
+
+			err = tx.Commit()
+			if err != nil {
+				d.Sync.Synced--
```
Let's add a rollback here.

WhoSoup replied:
just did!
